[SYCL] Have default_selector consider SYCL_BE

Alexander-Johnston opened the pull request:
Have the default_selector consider the env var SYCL_BE when rating
device scores to make choosing a backend easier.

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL] Have default_selector consider SYCL_BE

Have the default_selector consider the env var SYCL_BE when rating
device scores to make choosing a backend easier.

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

fwyzard commented on sycl/source/device_selector.cpp:
```diff
@@ -31,6 +31,24 @@ device device_selector::select_device() const {
+    if (DriverVersion.find("CUDA") == std::string::npos && backend == "PI_CUDA") {
+      return -1;
+    }
+    // We can't easily check for an OpenCL device as there is no common string
```
would it make sense to check for SPIR support, e.g. `dev.has_extension("cl_khr_spir")` ?

Alexander-Johnston replied:
We found we could reliably get a string containing OpenCL or CUDA from the platform information, rather than the device information, so won't have to rely on checking the extensions to determine OpenCL support.

Commit:
[SYCL] Select GlobalPlugin based on SYCL_BE

Rather than choose the last found plugin as GlobalPlugin, select
it depending on the SYCL_BE env var.

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

Commit:
[SYCL] Improve default device selection checks

Better checks for CUDA and OpenCL devices to match with SYCL_BE in the
default device selection, based on the platform version info.

Signed-off-by: Alexander Johnston <alexander@codeplay.com>

bjoernknafla commented on sycl/source/detail/pi.cpp:
```diff
@@ -196,11 +196,18 @@ vector_class<plugin> initialize() {
+        PluginNames[I].find("cuda") != std::string::npos) {
+      // Use the CUDA plugin as the GlobalPlugin
+      GlobalPlugin = std::make_shared<plugin>(PluginInformation);
+    }
```
I guess you are not using `if-else` here as we might support multiple active plugins in the future?

Should we then at least add an `assert(GlobalPlugin);` to "document" that a global plugin is required?

Alexander-Johnston replied:
I'm not sure the assert is needed. This function already fails if there is no backend selected or no plugin available.

I also think ideally we shouldn't need GlobalPlugin, and that it's counter to the multiple plugin design, but that's a larger discussion.

bjoernknafla commented on sycl/source/device_selector.cpp:
```diff
@@ -31,6 +31,25 @@ device device_selector::select_device() const {
+  // Take note of the SYCL_BE environment variable when doing default selection
+  const char *SYCL_BE = std::getenv("SYCL_BE");
+  std::string backend = (SYCL_BE ? SYCL_BE : "");
+  if (backend != "") {
```
If would be more efficient (no string comparison) to check SYCL_BE for being NULL - though probably does not matter much:

```suggestion
  if (SYCL_BE != nullptr) {
```

bjoernknafla reviewed (approved):
+1

Ruyk reviewed (approved):
Looks OK

Commit:
[SYCL] Formatting update for device_selector.cpp

Signed-off-by: Alexander Johnston <alexander@codeplay.com>
